lib/heuristics/scheduling_heuristic.rb

```diff
@@ -1086,22 +1062,23 @@ def prepare_output_and_collect_routes(vrp)
     def select_point(insertion_costs, empty_route)
       ### chose the most interesting point to insert according to [insertion_costs] ###
       if empty_route
-        if @routes_referents.empty?
+        if @candidate_routes.all?{ |_vehicle_id, all_routes| all_routes.all?{ |_day, route_data| route_data[:stops].empty? } }
           # chose closest with highest frequency
-          to_consider_set = insertion_costs.group_by{ |s| @services_data[s[:id]][:visits_number] }.sort_by{ |nb, _set| nb}.reverse.first[1]
+          to_consider_set = insertion_costs.group_by{ |s| @services_data[s[:id]][:raw].visits_number }.max_by{ |nb, _set| nb }[1]
           return to_consider_set.min_by{ |s| ((@services_data[s[:id]][:priority].to_f + 1) / @max_priority) * s[:additional_route_time] }
         else
           # chose distant service with highest frequency
           # max_priority + 1 so that priority never equal to max_priority and no multiplication by 0
-          return insertion_costs.max_by{ |s| (1 - (@services_data[s[:id]][:priority].to_f + 1) / @max_priority + 1) * @routes_referents.collect{ |ref| matrix(@candidate_routes[@candidate_routes.keys.first].first[1], ref, s[:point]) }.min * @services_data[s[:id]][:visits_number]**2 }
+          referents = @candidate_routes.collect{ |_vehicle_id, all_routes| all_routes.collect{ |_day, route_data| find_referent(route_data) } }.flatten.compact
+          return insertion_costs.max_by{ |s| (1 - (@services_data[s[:id]][:priority].to_f + 1) / @max_priority + 1) * distance_to_referents(referents, s).min * @services_data[s[:id]][:raw].visits_number**2 }
```

Comment:
this `distance_to_referents` function is used only here.  @fonsecadeline do you plan to use this array somewhere else?

Reply:
I could rename it to make it more global. `distance_to_set` would be a usefull function soon.

Reply:
I think the naming was okay, my comment was related to the fact that we create this array of distances to take the `min` and immediately discard the array. (this is why I asked if this array will be used somewhere else) IF it is not going to be useful somewhere else then instead of doing `collect` inside the `distance_to_referents` function and taking the `min` here, we can make the function so that it returns the `min` directly -- like `min_referent_distance` for example.

Reply:
First I thought it would be useful only here, but then I realized it will be useful for my potential next developments : for instance, we talked about trying to detect if a service is far from others. I do not know yet if I will want to compute average/min or max distance to others. So I guess this is a most generic function for now.

And I did not want to keep _referent_ in function name because the set can contain other services than only referents.